S[20] : As a user, I should be able to view the requests stream.

heba-aamer opened the pull request.

heba-aamer committed:
Issue#109-Adding-Requests-Stream

heba-aamer committed:
Issue#109 - Adding-Documentation

heba-aamer committed:
#109 - Adding-Filtering-Part

heba-aamer committed:
#109 -Renaming-Activities

heba-aamer committed:
#109 - Fixing-Redirections

heba-aamer committed:
Merge branch 'develop' of https://github.com/ndomar/megasoft-14 into Issue#109-View-Requests-Stream

heba-aamer committed:
#109 -Fixing-an-API-Issue

heba-aamer committed:
#109 - Fixing-Getting-Data-Issue

heba-aamer commented:
@Omar-ElAzazy , @mansoursaid001  please review the code.

MohamedBassem commented:
@heba-aamer , Here are my comments : 

1- Activity class names should be suffixed with "Activity" to differentiate between them and other classes.

2- Check inline comments.

3- Why do you use supressLint ?

4- I don't think that a dropdown list for users and tags is a good idea. If we have a tangle with thousands of tags the user will have to download a big amount of data also will have to search through this list for his tag which is not something good from the UX point of view. I believe that a text search will be a better idea. 

5- Also the search while typing thing is not a good idea. Let's say I'm search for "design", I typed "d" the server will go and search the whole database for all requests that contains the character "d" and send it to me which I'll have to download on my bundle , but I don't even need it as I was going to write "design". So I believe that there should be a search button instead of search while typing.

I'm impressed with the amount of details handled in the code :D Great work (Y)

Omar-ElAzazy commented:
@MohamedBassem Regarding the count element in the JSON file, that's a convention that has been used in all end-points implemented by me. That's basically a convention for any response with a list. It might turn to be useless in this case but in other cases it is easier for the end-point user to have it for the user to retrieve the count faster than explore the whole list to get the count. Still I agree that it might be useless in this part but that's being used because it is mainly exposed by the end-point I implemented.

Regarding the drop-down list, thinking about a tangle we assumed for this phase we won't have "thousands" of users and tags in a tangle which is more realistic than your assumption. 

Also for the search while typing and the drop-down list, that would be applied while searching for users or tags which makes more sense to have since as a user you will have to memorize the tag names to do it your way which is not a good thing for the user experience given that a word (especially tag names and user names) can be written in many different ways. And again, there shouldn't be "thousands" of users or tags per tangle anytime soon.

Omar-ElAzazy commented:
@MohamedBassem And regarding the amount of data sent during the creation of the drop-down list, that can be optimized. One way is to prevent any duplicated data sent/received. One way to do that is that whenever you enter a tangle you start receiving a list of users and tags in this tangle (You don't need to wait for this for the tangle to be viewed and other things to work so it doesn't make you wait) and whenever you start searching it sends asking for any additional users/tags in the system that were added since the last time you requested this list, just receive those added names and add them offline in your android app. And you maintain a data structure at the back-end of the Android application (for example a Trie) to search much faster in the list of users/tags. That's just a example how data usage can be optimized, but I don't think we really need anything of this since the list can't be that large anytime soon anyway.

MohamedBassem commented:
@Omar-ElAzazy , you won't explore the whole list it's just .getJSONArray().length(). But actually having it in the API is harmless so OK we can leave it.
- Having thousands of tags maybe it's a little bit unrealistic. Lets say 200 tags ( 40 requests \* 5 tags ). Should I search 200 tags whenever I need to search for a certain tag ? I don't think there's a website that gives you a dropdown list of its tags of something. And the user don't have to remember the exact tag name we will search using wildcards , for example "design" will get "designing" , "Blah design" , ... etc.
- Concerning the search while typing it's completely inefficient , with every character you type you send an HTTP request and execute a DB query for a result most probably you won't need. Also your speed of writing will be much more faster than waiting for the response while typing. so searching for "website designing request" 20 HTTP request will be sent back and forth with 20 DB queries executed but I don't need only except the last one.
- Concerning the amount of data in tag/user search , its not a big deal here I'm talking from the UX point of view as i mentioned above. But in the full text search , yes it is a big problem , you can't cache it here and most probably you will download the whole tangle requests if you just typed "a" while typing another word.

MohamedBassem commented:
@heba-aamer , another thing I forgot to mention

A bug that can be reproduced as follows :
- Filter by tags.
- Filter by user ( It will fail because the endpoint tangleId/{tangleId}/users is not implemented yet , or in the real case if the internet is disconnected ).
- Tags dropdown box is still there and whenever you choose any tag you will send a request with query parameters userid=XX which is the id of the tag.

I guess you shouldn't change the type of the current filter until the users/tags list is received.

Omar-ElAzazy commented:
@MohamedBassem Yes, that's when you are using Java, I was talking about the end-point convention in general. It can be removed though. That can be updated in the coming sprint to update all end-points doing it this way.

I think I know a website that does that, Twitter. I don't believe it is efficient to have all tags appearing and searching among them, but that's the first version of it, later on it should be more optimized, having maybe some time lag among requests to allow the user to enter more information about the search (i.e. complete the word (s)he is searching for). we have other ideas in mind for the search but since this sprint was one week long we decided to do it this way for now and improve it later on adding other functionalities. (i.e. using wild cards is one of the ideas I had in mind, having an advanced search page is another idea).  But still having a drop-down list is not a bad idea, actually it would help the user and improve the UX.

I don't think that there is searching while typing in the full-text search. It only exists while searching for users/tags. Or at least that's how it should be.

MohamedBassem commented:
@Omar-ElAzazy , there is no search while typing in tags/users because there isn't any typing , you just choose from the dropdown list. The search while typing thing is in the full text search.
Still having the 200 tag in the dropdown box isn't a good idea to scroll in ( No typing here ).

Omar-ElAzazy commented:
@MohamedBassem The target of the drop-down list is to have it while typing, that should be more apparent after next sprint, consider this the first version of what will be a search while typing.

MohamedBassem commented:
@Omar-ElAzazy , Aha , if it's like that I agree. But I completely disagree with the fulltext search while typing :D

heba-aamer commented:
@MohamedBassem , here are my answers:
- savedInstanceState Bundle:    
  - you are right in my code i have removed it two days ago but i did not pushed it after changing it , and the same goes for the SuppressLint.
- Crashing problem:
  - fixed.
- The ActivityName:  
  - The name of the activity is the same as proposed by Eslam's mockups that we should follow. anyway i agree on putting Activity as suffix but this will be a convention for the whole company that we should settle it down before the beginning of the next sprint.
- API count:
  - answered by Omar.
- Fragment:
  - i started reading about Fragments but i did not realize that it could be that specific. i will see how it could be done for the coming sprint. Thanks for the info.
- Headers:
  - There are 2 types of headers "X-SESSION-ID" for the tag end point, while the other header "SessionId" was for the user end point. (they do not have the same structure).
- The bug that could be reproduced:
  - First of all, there is an implemented end point to get all the users (GET /tangle/{tangleId}/user), it was /users before the deadline but i checked now it was changed to /user so i modified the related part.
  -  i think before sending any request we should check first if there is an internet connection or not and this will handle this case.
  - but i do not get, how there is no connection and i will be able to send the request with tag value for the userid :D.
- Full text search:
  - i totally agree that this was not efficient, by the way adding a button will be much easier than using the TextWatcher :).  
- Drop down list Tag/User:
  - as Omar said , this will be enhanced in the coming sprint isA.

Thanks for the review.

MohamedBassem commented:
@heba-aamer ,
- Activity Name : I told Eslam and he told me that he will fix it. I didn't know about the mockups btw :D So most probably I'll change mine too.
- Headers , all the headers in the API should have the same structure.
- The bug that could be reproduced , if anything happened to the HTTP request , the server was busy or something so it didn't respond to the request.  You can send the userid once you get your internet connection back :D And yes we should handle the internet connection status in the next sprint or something.
- Full text search , Yes I know , that's why I think a button will be much better even in the implementation.

MohamedBassem commented:
Btw in your last commit , you committed files that shouldn't be committed.

heba-aamer committed:
#109 add the files again
